### PR TITLE
[H/3] Wait for `WritesClosed` and `Dispose` as fire-and-forget

### LINKED ISSUE #77139: HTTP/3 connection or stream shutdown is slow
### Description

gRPC HTTP/3 benchmark performance drastically decreased. It looks like something with stream/connection shutdown is taking forever, which affects RPS calculation. (And slow shutdown doesn't seem like a good thing regardless of RPS calculations)

### Configuration

Linux citrine perf machine, crank commands below

### Regression?

Yes, something in the following changes caused HTTP/3 gRPC tests to be really slow (shutdown takes forever which causes the RPS calculation to be longer)
https://github.com/dotnet/runtime/compare/b7fca07f478739d9fd50bdf5d1467c02dc05b6b0...1cb2f42e1e552e03895ad032b0b03949eec44c4f

### Data

Changing **only** the runtime version used by the client shows a difference in RPS.

### Before the changes:
~3.8m RPS
<details>
<summary>crank command</summary>

```
crank --config https://raw.githubusercontent.com/aspnet/Benchmarks/main/build/ci.profile.yml --config https://raw.githubusercontent.com/aspnet/Benchmarks/main/scenarios/grpc.benchmarks.yml --scenario grpcaspnetcoreserver-grpcnetclient --profile intel-lin-app --profile intel-load-load --variable scenario=serverstreaming --variable requestSize=0 --variable responseSize=0 --variable streams=70 --variable connections=20 --variable threads=20 --variable protocol=h3 --application.framework net7.0 --load.framework net7.0 --application.options.collectCounters true --load.runtimeVersion 7.0.0-preview.7.22362.12
```

</details>

### After the changes:
~300k RPS
<details>
<summary>crank command</summary>

```
crank --config https://raw.githubusercontent.com/aspnet/Benchmarks/main/build/ci.profile.yml --config https://raw.githubusercontent.com/aspnet/Benchmarks/main/scenarios/grpc.benchmarks.yml --scenario grpcaspnetcoreserver-grpcnetclient --profile intel-lin-app --profile intel-load-load --variable scenario=serverstreaming --variable requestSize=0 --variable responseSize=0 --variable streams=70 --variable connections=20 --variable threads=20 --variable protocol=h3 --application.framework net7.0 --load.framework net7.0 --application.options.collectCounters true --load.runtimeVersion 7.0.0-rc.1.22363.10
```

</details>

### Analysis

I narrowed down the changes in runtime to https://github.com/dotnet/runtime/compare/b7fca07f478739d9fd50bdf5d1467c02dc05b6b0...1cb2f42e1e552e03895ad032b0b03949eec44c4f which contains some Quic API changes that may be involved in the regression.

### PR DESCRIPTION
**Pull request by ManickaP**
Alternative to #104032. Solves the perf problem for waiting `Dispose` (with sync-over-async) by waiting on `QuicStream.WritesClosed` at the end of content writing together with doing `QuicStream` disposal as fire-and-forget in case the writes are already closed (avoiding potential data-loss).

cc @liveans 

Fixes #77139

### CODE REVIEW

**Comment by liveans** [comment]
Related to #102571, let's run this on http benchmarks.

**Comment by ManickaP** [comment]
I just ran the gRPC benchmarks on this:
- 9.0 - base line
```
| load                    |                                       |
| ----------------------- | ------------------------------------- |
| Max CPU Usage (%)       | 73                                    |
| Max Cores usage (%)     | 2,037                                 |
| Max Working Set (MB)    | 460                                   |
| Max Private Memory (MB) | 2,578                                 |
| Build Time (ms)         | 12,851                                |
| Start Time (ms)         | 100                                   |
| Published Size (KB)     | 86,443                                |
| Symbols Size (KB)       | 176                                   |
| .NET Core SDK Version   | 9.0.100-preview.7.24323.5             |
| ASP.NET Core Version    | 9.0.0-preview.6.24320.4+613c1e990b6b  |
| .NET Runtime Version    | 9.0.0-preview.6.24319.11+117cfccdd71a |
| Server GC enabled       | True                                  |
| Processor Count         | 28                                    |
| First Request (ms)      | 1,482.92                              |
| Requests/sec            | 966,263                               |
| Requests                | 19,634,472                            |
| Bad responses           | 0                                     |
| Mean latency (ms)       | 0.77                                  |
| Max latency (ms)        | 15,285.50                             |
```
- this PR
```
| load                    |                                       |
| ----------------------- | ------------------------------------- |
| Max CPU Usage (%)       | 73                                    |
| Max Cores usage (%)     | 2,033                                 |
| Max Working Set (MB)    | 433                                   |
| Max Private Memory (MB) | 1,209                                 |
| Build Time (ms)         | 7,754                                 |
| Start Time (ms)         | 101                                   |
| Published Size (KB)     | 86,443                                |
| Symbols Size (KB)       | 176                                   |
| .NET Core SDK Version   | 9.0.100-preview.7.24323.5             |
| ASP.NET Core Version    | 9.0.0-preview.6.24320.4+613c1e990b6b  |
| .NET Runtime Version    | 9.0.0-preview.6.24319.11+117cfccdd71a |
| Server GC enabled       | True                                  |
| Processor Count         | 28                                    |
| First Request (ms)      | 1,597.35                              |
| Requests/sec            | 3,850,558                             |
| Requests                | 20,011,352                            |
| Bad responses           | 0                                     |
| Mean latency (ms)       | 0.41                                  |
| Max latency (ms)        | 5,059.72                              |
```

The base line 966,263 RPS is not as bad as reported in #77139 (which was meager ~300k RPS). But this PR with 3,850,558 RPS  is at the same numbers as before the issue was reported ~3.8m RPS

**Review by liveans — APPROVED** [review]
Thanks!

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs [hunk 1]
```diff
@@ -88,7 +88,14 @@ public void Dispose()
             {
                 _disposed = true;
                 AbortStream();
-                _stream.Dispose();
+                if (_stream.WritesClosed.IsCompleted)
```
Comment (MihaZupan) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:91]: Is `_stream.WritesClosed.IsCompleted` expected to be true most of the time for basic GET/POST requests (when we're not doing bidirectional streaming)?

Comment (ManickaP) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:91]: Yes.

**Review by rzikm — APPROVED** [review]
LGTM if CI is green.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs [hunk 2]
```diff
@@ -512,6 +527,10 @@ private async ValueTask FlushSendBufferAsync(bool endStream, CancellationToken c
         {
             await _stream.WriteAsync(_sendBuffer.ActiveMemory, endStream, cancellationToken).ConfigureAwait(false);
             _sendBuffer.Discard(_sendBuffer.ActiveLength);
+            if (endStream)
+            {
+                await _stream.WritesClosed.ConfigureAwait(false);
+            }
 
             await _stream.FlushAsync(cancellationToken).ConfigureAwait(false);
```
Comment (MihaZupan) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:535]: It feels odd that we're flushing after we've ensured that writes are complete. If the flush is a no-op can we delete it?

Comment (ManickaP) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:535]: I'll switch the order.

**Pull request merged**